feat: Waypoint management GUI

ShadowCat117 opened the pull request:
All works as far as I can tell. No import/export feature or filtering by icon

ShadowCat117 committed:
feat: Base implementation of waypoint GUI

Edit and Delete buttons render, but do not work

ShadowCat117 committed:
fix: Added manager icon

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiCreationScreen.java:
```diff
@@ -65,7 +65,8 @@ public final class PoiCreationScreen extends WynntilsScreen implements TextboxSc
 
-    private final MainMapScreen oldMapScreen;
+    private MainMapScreen oldMapScreen;
+    private PoiManagementScreen managementScreen;
```
You can just merge this and save these as `Screen`

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -1293,6 +1293,21 @@
+  "screens.wynntils.map.manager.name": "Manage Waypoints",
+  "screens.wynntils.map.manager.description1_1":  "List, Delete or Create",
+  "screens.wynntils.map.manager.description1_2":  " all your preview set",
+  "screens.wynntils.map.manager.description1_3":  " waypoints!",
```
You can just make these description1, description2, description3, as there are no subcategories

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -1293,6 +1293,21 @@
+  "screens.wynntils.map.manager.description1_3":  " waypoints!",
+  "screens.wynntils.poiManagementGui.close":  "X",
+  "screens.wynntils.poiManagementGui.next":  ">",
+  "screens.wynntils.poiManagementGui.previous":  "<",
```
No need to be excessive with translations. These can be made literals, they will never change in any language.

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -1293,6 +1293,21 @@
+  "screens.wynntils.poiManagementGui.name":  "Name",
+  "screens.wynntils.poiManagementGui.x":  "X",
+  "screens.wynntils.poiManagementGui.y":  "Y",
+  "screens.wynntils.poiManagementGui.z":  "Z",
```
I would remove these from here too

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,132 @@
+        float centreZ = 64 + spacingMultiplier * row;
+
+        RenderUtils.drawTexturedRect(
+                poseStack, POI_ICONS.get(POI_ICONS.indexOf(poi.getIcon())), this.width / 2f - 151 - (poi.getIcon().width() / 2f), centreZ - (poi.getIcon().height() / 2f));
```
`poi.getIcon()` returns the texture? Why do we look up the index and get it?

kristofbolyai replied:
(Don't forget to remove `POI_ICONS` when you simplify this)

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,132 @@
+    private Button editButton;
+    private Button deleteButton;
+    private int row;
+    private int colour;
```
```suggestion
    private CustomColor color;
```

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,132 @@
+    private Button deleteButton;
+    private int row;
+    private int colour;
+    private int spacingMultiplier = 20; //Small 14, medium 20, large 25
```
Are these planned to be changeable or what is the purpose?

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,132 @@
+        System.out.println("X: " + this.getX() + "\nY: " + this.getY() + "\nWidth: " + this.width + "\nHeight: " + this.height);
+    }
+
+    private void renderIcon(PoseStack poseStack, CustomPoi poi, int row) {
```
What is stopping you from using `IconPoi#renderAt`?

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,132 @@
+            colour = 0x636363;
+        }
+
+        int groupShift = group == ungroupedIndex ? 20 : 0;
```
Honestly I have no idea what any of these 3 variables do

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,199 @@
+    private List<CustomPoi> waypoints;
+    private int pageHeight;
+    private int page;
+    private int spacingMultiplier = 20; //Small 14, medium 20, large 25. Make customisable
```
Left a comment in the widget itself but I don't think this is a good idea. If we make this a constant 20, does it not work on a specific GUI scale?

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,199 @@
+    @Override
+    public void doRender(PoseStack poseStack, int mouseX, int mouseY, float partialTick) {
+        renderBackground(poseStack);
+        super.doRender(poseStack, mouseX, mouseY, partialTick);
```
Does this super call do anything (same goes for `super.doInit`)?

kristofbolyai replied:
Sorry, this is likely used to render your widgets, when you register them. However, the question still stands for doInit.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,199 @@
+                    this));
+        }
+
+        poiManagerWidgets.forEach(widget -> widget.render(poseStack, mouseX, mouseY, partialTick));
```
Doing anything like adding widgets during render time is strictly bad design (legacy :)). Make this a method, do this on init and whenever we delete (edit will reopen this screen so init will handle it)

kristofbolyai replied:
No wonder why your mouse clicks are not getting registered. I do not know how legacy did this, but here you need to use `addRenderableWidget` to register for widgets to be rendered, and all events will also be called automatically (like mouse events). Use `removeWidget` to clear widgets using your `poiManagerWidgets`.

ShadowCat117 committed:
feat: Added functionality to edit and delete buttons

Also includes up/down buttons to reorder poi's in pages and other improvements made based on review

ShadowCat117 committed:
Scrolling now acts like next/previous buttons

Also removed some unused imports

ShadowCat117 committed:
Removed groupShift variable as it seemed to always be 20 anyway

Not 100% sure but I assume it was used in legacy for something with the group filter which Artemis doesn't have. Seems to work fine without it

ShadowCat117 committed:
Added delete all button

ShadowCat117 committed:
fix: Delete button was taking 1 extra click

ShadowCat117 committed:
Added line underneath page headers

ShadowCat117 committed:
Removed leftover print

ShadowCat117 commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,253 @@
+                    0,
+                    0,
+                    this.width,
+                    this.height,
```
This is the only part I'm unsure of if it's done well, as far as I can tell the widget renders fine but I would prefer it with actual values.

I did try inputting what I thought was correct but it then ended up moving each row to the very left of the window but the Y value was correct so seemingly the X was wrong which I was trying with `this.width/2 - 151` which is the top left position of the icon as that's the leftmost part of the widget.

I've tried resizing my window and different GUI scales and it looks good so it doesn't seem to be a problem but if there's a scenario where it would be a problem then I'm not certain on how to do this correctly

ShadowCat117 commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,180 @@
+
+        float centreZ = 64 + spacingMultiplier * row;
+
+        poi.renderAt(poseStack, bufferSource, this.width / 2f - 151, centreZ, false, 1f, 1f);
```
When there is only 1 widget being rendered on the page this doesn't display the icon.

magicus replied:
Huh? This comment seems out of place. What was the purpose of it?

ShadowCat117 replied:
Sorry let me be more clear. When a page is only displaying 1 poi, so the final page, the icon for that poi doesn't render. This code is definetly being ran as the name and all other information about the poi render but for some reason the icon doesn't.
![2023-03-30_13 05 37](https://user-images.githubusercontent.com/8337467/228830815-6a062280-74fc-4e65-a2d0-c0d4274c296d.png)

magicus replied:
Ok, now I understand: you have a bug that you cannot resolve, and need help fixing it.

magicus replied:
Nothing really sticks out as being wrong. The one thing I react to is that in this case `row` is 0, so you don't get `spacingMultiplier` in work at all. Maybe some maths is wrong, so it is technically being rendered, but off-screen? Maybe try this hypothesis by offsetting row with one? Then all icons should be rendered one step down.

ShadowCat117 replied:
That sounds like it could be it, I'll give that a try later

ShadowCat117 replied:
Well I know that's not the problem but I do know what is. It happens when all icons on the same page are the same type.
![2023-03-30_18 35 21](https://user-images.githubusercontent.com/8337467/228919415-ffd3c6fd-8864-45a1-a076-1444caf4fdc8.png)

magicus replied:
There might be a problem with the poi.renderAt() method. Hm, interesting...

magicus replied:
I don't think you can use poi.renderAt() here. It is not supposed to work like that. For instance, it takes the mapZoom value and checks the transparency of the icon at that zoom level.

magicus replied:
You are hard-coding a zoom level of 1f, I guess T3 loot chests are invisible at that zoom level...

kristofbolyai replied:
> There might be a problem with the poi.renderAt() method. Hm, interesting...

I suggested that.. We used it at world distance feature. Maybe it works differently..

magicus replied:
I'm not sure what a good solution would be. Either add a renderForGui() in Poi.java, or if your pois are only IconPoi's, call getIcon() and render it yourself.

kristofbolyai replied:
> I'm not sure what a good solution would be. Either add a renderForGui() in Poi.java, or if your pois are only IconPoi's, call getIcon() and render it yourself.

You have to handle colors, so method in IconPoi.

ShadowCat117 replied:
For now I've just changed it back to using the same icon rendering method that PoiCreationScreen uses. I did try making a renderForGui in IconPoi where the alpha was always 1 and setting z to 1 for all pois but it just displayed exactly the same as renderAt.

If this is acceptable for now and you don't see any other areas I can improve I guess this'll be ok. Also not sure if you've seen the comment I have above this one, I'm not 100% sure if that's ok to have the way it is

kristofbolyai replied:
Looks good in-game now.

ShadowCat117 committed:
fix: Deleting final poi on a page now goes to previous page

kristofbolyai commented:
@JamieCallan117 If this is ready, please mark it as "ready for review". Also, if you think you have completed a request, please mark the conversation as resolved.

Commit:
ci: spotless formatting

ShadowCat117 committed:
Merge branch 'main' into PoiManagerGui

magicus commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -1311,6 +1311,10 @@
   "screens.wynntils.map.help.description9": "Shift-Right click to center on the compass marker.",
   "screens.wynntils.map.help.name": "Quick Guide",
+  "screens.wynntils.map.manager.description1": "List, Delete or Create",
+  "screens.wynntils.map.manager.description2": " all your preview set",
```
This does not really make sense?

ShadowCat117 replied:
I just copied it from legacy originally and noticed that too, just forgot to change it to something better

magicus commented on common/src/main/java/com/wynntils/screens/maps/MainMapScreen.java:
```diff
@@ -99,6 +99,28 @@ protected void doInit() {
+                        Component.literal("[>] ")
+                                .withStyle(ChatFormatting.RED)
+                                .append(Component.translatable("screens.wynntils.map.manager.name")),
+                        Component.translatable("screens.wynntils.map.manager.description1")
```
Did you intend for the descriptions to come on separate lines? Then you have to send them as separate item in the list. By using append they will all become a single line -- this may be fine, but if that is the intention you can have just a single i18n key and put it all on a single line.

kristofbolyai replied:
While this sounds logical, it might not be the case, I am pretty sure this is a copy of what I did above, and it works.

ShadowCat117 replied:
Yes, I just copied what the other buttons did and it works
![2023-03-30_13 11 26](https://user-images.githubusercontent.com/8337467/228832087-4c7d4b6f-9bb7-4219-93f9-7e0b895b7bf9.png)

magicus replied:
Okay but then you don't need three different i18n keys, you can just have a single `description` key.

magicus commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright © Wynntils 2022.
```
It's 2023 now...

magicus commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+                        .size(20, 20)
+                        .build());
+
+        this.addRenderableWidget(
```
This sounds very dangerous and drastic. Are we sure this is even needed? If someone mess up their waypoints that much, it is probably better to edit the json config files.

ShadowCat117 replied:
It does have to be pressed 3 times so while dangerous I think most people will not accidently trigger it. It was a feature in legacy so I ported it, but if you think it's not worth having then I can easily remove it

magicus replied:
Just because it was that way in legacy does not mean it was a good feature. :-) Apart from the risk of e.g. an autoclicker getting astray, there is also the risk that we somehow ourselves call such a function. I have a hard time seeing how this could ever be useful, compared to the risk that something goes wrong. 

Let's hear @kristofbolyai's opinion, but otherwise I'd recommend that you remove it, yes.

kristofbolyai replied:
You can reset pois via a command, if you really want to. So I don't think we need this.

magicus commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+        for (int i = 0, lim = Math.min(pageHeight, waypoints.size() - pageHeight * page); i < lim; i++) {
+            CustomPoi poi = waypoints.get(page * pageHeight + i);
+
+            if (poi == null) {
```
How can this happen?

ShadowCat117 replied:
In theory it shouldn't, it's just a safety precaution in case it somehow does. Can remove it if you think it's pointless having it but I just thought better to be on the safe side

magicus replied:
I think it is pointless. Even if this happened and you skipped one poi, the index would be out of phase and stuff will likely break anyway. If you get a NPE, we will register this as a crashed screen and send a report, so we can investigate.

kristofbolyai replied:
Why did you resolve this, but not remove it?

ShadowCat117 replied:
I have removed it?

https://github.com/Wynntils/Artemis/pull/1382/commits/ff6e943cb14eedd26a690b5d88abf33027b0105a#diff-908a218ec807dddcff8aababd8ececfbd7171a996e737521ef43a7fffb7b9ea6L182

kristofbolyai replied:
@JamieCallan117 Sorry, my github was high..

magicus commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,174 @@
+        color = CustomColor.fromInt(0xFFFFFF);
+
+        if (poi.getVisibility() == CustomPoi.Visibility.HIDDEN) {
+            color = CustomColor.fromInt(0x636363);
```
Please extract magic numbers as a constant

kristofbolyai replied:
CommonColors.GRAY would be fine, no need to have this specifically set

magicus commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,174 @@
+
+        pois = Managers.Feature.getFeatureInstance(MapFeature.class).customPois.get();
+
+        color = CustomColor.fromInt(0xFFFFFF);
```
We have CommonColors.WHITE for this purpose

kristofbolyai replied:
Here, and below.. resolved but not changed?

magicus commented on common/src/main/java/com/wynntils/screens/maps/widgets/PoiManagerWidget.java:
```diff
@@ -0,0 +1,174 @@
+
+        this.deleteButton = new Button.Builder(
+                        Component.translatable("screens.wynntils.poiManagementGui.delete"), (button) -> {
+                            Managers.Feature.getFeatureInstance(MapFeature.class)
```
Having a delete without either a confirmation, or a way to undo, is a bit unforgiving.

(I'd prefer a "Undo last delete" rather than a confirmation)

ShadowCat117 replied:
I could make the delete button require 2 presses, similar to the delete all button, as well as the undo last delete, or would that be overkill?

magicus replied:
I think that's overkill. Look at how Github does when you delete a branch after merging a PR. I think that's pretty nifty. You can delete it with just a single click, but if you mis-clicked, you get the chance to undo. So maybe do something like this:
when deleting a poi, save it as "last deleted" in the screen, and in the screen, if there is a "last deleted" poi, show a "undo delete" button (which is otherwise hidden, or possibly visible but greyed out), and if that is pressed, just re-add the poi. And don't bother trying to persist the "last deleted poi", if you exit the screen well then you're out of luck.

I think that'd work nicely.

ShadowCat117 replied:
I've added the undo last delete button, but it only stores 1 deletion, do you think it would be a good idea to instead store a list of deleted pois so then you can undo everything you've deleted in the current "session".

magicus replied:
No, I think that is enough. It's a bit of a safety, but we're not overdoing it.

kristofbolyai commented on common/src/main/java/com/wynntils/utils/render/Texture.java:
```diff
@@ -86,6 +86,7 @@ public enum Texture {
     MAP_HELP_BUTTON("map/map_help_button.png", 10, 16),
     MAP_POINTERS("map/map_pointers.png", 256, 256),
     MAP_SHARE_BUTTON("map/map_share_button.png", 16, 14),
+    MAP_MANAGER_BUTTON("map/map_manager_button.png", 12, 16),
```
Please keep these groups sorted.

ShadowCat117 replied:
Is it meant to be alphabetical? Because if so these should be swapped too
`MAP_DEFENSE_FILTER_BUTTON("map/map_defense_filter_button.png", 16, 16),
     MAP_BUTTONS_BACKGROUND("map/map_buttons_background.png", 144, 26),`

kristofbolyai replied:
Yes, please sort that too. It was overlooked in the PR it was added in.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+        waypoints =
+                Managers.Feature.getFeatureInstance(MapFeature.class).customPois.get();
+
+        if (!waypoints.isEmpty() && page * pageHeight > waypoints.size() - 1) {
```
Checking `waypoints.isEmpty()` seems pretty redundant

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+
+        RenderUtils.drawRect(poseStack, CommonColors.WHITE, this.width / 2 - 165, 52, 0, 355, 1);
+
+        poiManagerWidgets.forEach(widget -> widget.render(poseStack, mouseX, mouseY, partialTick));
```
You don't need this as you call super.render.

ShadowCat117 replied:
I've tried that but it seems to make some icons randomly dissappear 

https://user-images.githubusercontent.com/8337467/228920141-9e7d07c0-902a-419e-9846-34b7bfa838bb.mp4

kristofbolyai replied:
Then, there is something really wrong..

magicus replied:
Yes, that is worrisome. Overall, the PR looks pretty okay now, but the root cause of the disappearing icons must be found.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+            previousPage();
+        }
+
+        for (int i = 0, lim = Math.min(pageHeight, waypoints.size() - pageHeight * page); i < lim; i++) {
```
Why set lim to a variable? Just inline it in the comparasion

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+        return true;
+    }
+
+    public void populatePois() {
```
Shouln't you update `waypoints` here to the new value?

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+        this.poiManagerWidgets.clear();
+
+        if (Math.min(pageHeight, waypoints.size() - pageHeight * page) == 0 && page != 0) {
+            previousPage();
```
Shouldn't you `page` to `waypoints.size() / pageHeight` to make sure you are on the last page, properly?

ShadowCat117 replied:
I don't think so, `waypoints.size() - pageHeight * page` calculates how many more pois there are to be displayed on this and the following pages. So with my GUI scale it displays 13 per page and if my final page only has 2 pois on the penultimate page then it the value will be 15 and on the final page it will be 2.

So if I deleted the 2 elements on the final page the min of those 2 values would be 0 and it will send you to the previous (now last) page.

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+    private int deleteCount = 3;
+    private List<CustomPoi> waypoints;
+    private int pageHeight;
+    private int page;
```
Rename this to `currentPage`

kristofbolyai commented on common/src/main/java/com/wynntils/screens/maps/PoiManagementScreen.java:
```diff
@@ -0,0 +1,243 @@
+    private List<CustomPoi> waypoints;
+    private int pageHeight;
+    private int page;
+    private int spacingMultiplier = 20;
```
If you say this value works for all GUI scales, then inline it.

ShadowCat117 committed:
Merge branch 'main' into PoiManagerGui

ShadowCat117 committed:
Changed manager button description on map screen

ShadowCat117 committed:
Removed delete all button

ShadowCat117 committed:
Bunch of minor code improvements

ShadowCat117 committed:
feat: Added button to undo last deletion

Commit:
ci: spotless formatting

ShadowCat117 committed:
Undo delete now stores more than 1 deleted poi

ShadowCat117 committed:
Merge branch 'PoiManagerGui' of https://github.com/JamieCallan117/Artemis into PoiManagerGui

ShadowCat117 committed:
Changed poi icon rendering

Commit:
ci: spotless formatting

ShadowCat117 committed:
Corrected order of 2 textures

kristofbolyai commented:
Pressing down button crashes. That down button should also be disabled.
![image](https://user-images.githubusercontent.com/49001742/229148578-d6a518ee-f545-4587-8e9a-ea16e35e2824.png)

ShadowCat117 committed:
fix: Down button disabled when only 1 poi exists

kristofbolyai committed:
Merge branch 'main' into PoiManagerGui

kristofbolyai commented:
I have no idea what happened with the merges.. Seems like you are out of sync with main and the build fails. Please fix it.

kristofbolyai reviewed (approved):
LGTM

ShadowCat117 commented:
Looks like my gradle is updating, that might be the issue, didn't let it finish beforming commiting those changes

ShadowCat117 committed:
fix: Updated alignment enums

magicus committed:
Merge branch 'main' into PoiManagerGui

magicus commented:
> Pressing down button crashes. That down button should also be disabled. ![image](https://user-images.githubusercontent.com/49001742/229148578-d6a518ee-f545-4587-8e9a-ea16e35e2824.png)

Was this resolved? It looked okay in the code, but I'd like to get it confirmed.

magicus reviewed (approved):
Code looks good. Thanks @JamieCallan117 for your contribution, and for having the patience of polishing it to match our code standards and requirements! This will be a welcome addition to Artemis.

ShadowCat117 commented:
> > Pressing down button crashes. That down button should also be disabled. ![image](https://user-images.githubusercontent.com/49001742/229148578-d6a518ee-f545-4587-8e9a-ea16e35e2824.png)
> 
> Was this resolved? It looked okay in the code, but I'd like to get it confirmed.
![2023-03-31_22 51 51](https://user-images.githubusercontent.com/8337467/229239753-55966352-55fe-4f2a-8aa1-2dc132615882.png)

Yes!

ShadowCat117 commented:
> Code looks good. Thanks @JamieCallan117 for your contribution, and for having the patience of polishing it to match our code standards and requirements! This will be a welcome addition to Artemis.

No problem! Honestly should be thanking you for reviewing all my code. It's my first time contributing to an open source project and a Minecraft mod so a lot of simple mistakes, hopefully if I add anything else it won't have so many

magicus commented:
I'm glad you've started your learning curve towards contributing to open source project. You're more than welcome to keep adding functionality to Wynncraft in the future!
